### DISCUSSION TITLE
Does WeakReference create a DestroyGCHandle event on garbage collection?

### DISCUSSION THREAD

**Discussion by interwhy** [dotnet/runtime#62310] [Q&A] [answered]
Hi, I'm investigating a GCHandle leak in an application and I'm having trouble finding documentation in this area. I believe I narrowed the source down through dotnet-trace and perfview to WeakReferences being created causing a SetGCHandle event without any corresponding DestroyGCHandle events, but someone else pointed out that it's not clear if WeakReferences would cause DestroyGCHandle events if they were garbage collected in gen 0.

I found these methods:
https://github.com/dotnet/runtime/blob/b83fee9f6e3442b9808eee3c436499c6f13a648e/src/coreclr/vm/gchandleutilities.h#L285-L298
and some of them don't have any references, such as DestroyWeakHandle. Does that mean the DestroyGCHandle event won't be created? I'm not too familiar with the inner workings of the .NET garbage collector, so any help or guidance would be huge for me 😄

**Comment by danmoseley** [comment]
@Maoni0

**Comment by Maoni0** [comment, marked as answer]
weakrefs should emit a DestroyHandle event when their finalizer is run in [`GCToEEInterface::EagerFinalized`](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/gcenv.ee.cpp#L1091). this will call [`DestroyTypedHandle`](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/weakreferencenative.cpp#L592).

**Reply by interwhy** [reply]
That answers my question, thank you!
